File: docs/manual/java/guide/cluster/code/docs/home/persistence/ChildActors.java
package docs.home.persistence;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.cluster.sharding.ClusterSharding;
import akka.cluster.sharding.ClusterShardingSettings;
import akka.cluster.sharding.ShardRegion;
import akka.testkit.javadsl.TestKit;
import akka.util.Timeout;
import com.lightbend.lagom.javadsl.persistence.ChildPersistentEntity;
import com.lightbend.lagom.javadsl.persistence.ChildPersistentEntityFactory;
import com.lightbend.lagom.javadsl.persistence.PersistentEntity;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import play.inject.Injector;

import javax.inject.Inject;
import java.util.Optional;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;

public interface ChildActors {

    //#create-child-persistent-entity-factory
    public class MyProcessManagerRouter {

        private final ActorRef processManager;

        @Inject
        public MyProcessManagerRouter(ActorSystem system, Injector injector) {
            ChildPersistentEntityFactory<MyCommand> factory =
                ChildPersistentEntityFactory.forEntity(
                    MyProcessManagerEntity.class, injector);

            processManager = ClusterSharding.get(system).start(
                "MyProcessManager",
                Props.create(() -> new MyProcessManager(factory)),
                ClusterShardingSettings.create(system),
                messageExtractor
            );
        }

        // ...
        //#create-child-persistent-entity-factory

        private ShardRegion.MessageExtractor messageExtractor;
    }

    public interface MyCommand {}

    public class StartProcess implements MyCommand, PersistentEntity.ReplyType<String> {

    }

    public class MyProcessManagerEntity extends PersistentEntity<MyCommand, String, String> {
        @Override
        public Behavior initialBehavior(Optional<String> snapshotState) {
            return newBehaviorBuilder("").build();
        }
    }

    //#create-child-persistent-entity
    public class MyProcessManager extends AbstractActor {

        private final ChildPersistentEntityFactory<MyCommand> factory;
        private ChildPersistentEntity<MyCommand> entity;

        public MyProcessManager(ChildPersistentEntityFactory<MyCommand> factory) {
            this.factory = factory;
        }

        @Override
        public void preStart() throws Exception {
            entity = factory.create(
                context().self().path().name(), // The entity id
                "entity",             // Name of the child actor
                context()                       // Actor context
            );
        }

        // ...
        //#create-child-persistent-entity

        @Override
        public Receive createReceive() {
            return receiveBuilder().build();
        }

        private void demonstrateTell() {
            //#child-persistent-entity-tell
            entity.tell(new StartProcess(), self());
            //#child-persistent-entity-tell
        }

        private void demonstrateForward() {
            //#child-persistent-entity-forward
            entity.forward(new StartProcess(), context());
            //#child-persistent-entity-forward
        }

        private void demonstrateAsk() {
            //#child-persistent-entity-ask
            CompletionStage<MappedReply> result =
                entity.ask(new StartProcess(),
                    Timeout.apply(3, TimeUnit.SECONDS)
                ).thenApply(MappedReply::new);

            akka.pattern.PatternsCS.pipe(result, context().dispatcher())
                .to(self());
            //#child-persistent-entity-ask
        }
    }

    public class MappedReply {
        public MappedReply(String reply) {}
    }

    // Because we don't want junit actually picking this up
    public @interface Test {}

    //#mock-child-persistent-entity
    public class MyProcessManagerTest {

        static ActorSystem system;

        @BeforeClass
        public void setup() {
            system = ActorSystem.create();
        }

        @AfterClass
        public void tearDown() {
            system.terminate();
        }

        @Test
        public void testStartProcess() {
            new TestKit(system) {{
                TestKit probe = new TestKit(system);

                ActorRef processManager = system.actorOf(
                    Props.create(() -> new MyProcessManager(
                        // Mock the child persistent entity factory to use
                        // our test probe.
                        ChildPersistentEntityFactory.mocked(
                            MyProcessManagerEntity.class,
                            probe.getRef()
                        )
                    ))
                );

                // Send the process manager a start message
                processManager.tell("start", getRef());
                // Expect the entity to receive a start process message
                probe.expectMsg(new StartProcess());
                // Simulate the entity to reply with a started message
                probe.reply("started");
                // Expect that message to be mapped forward back to us
                expectMsg(new MappedReply("started"));
            }};
        }
    }
    //#mock-child-persistent-entity

}

File: docs/manual/scala/guide/cluster/code/ChildActors.scala
package docs.scaladsl.cluster.childactors

import akka.actor.{Actor, ActorSystem, Props}
import akka.cluster.sharding.{ClusterSharding, ClusterShardingSettings, ShardRegion}
import akka.testkit.{TestKit, TestProbe}
import akka.util.Timeout
import com.lightbend.lagom.scaladsl.persistence.PersistentEntity.ReplyType
import com.lightbend.lagom.scaladsl.persistence.{ChildPersistentEntity, ChildPersistentEntityFactory, PersistentEntity}
import org.scalatest.{BeforeAndAfterAll, WordSpecLike}

//#create-child-persistent-entity-factory
class MyProcessManagerRouter(system: ActorSystem, entityFactory: () => MyProcessManagerEntity) {
  private val processManager = {
    val factory = ChildPersistentEntityFactory.forEntity(entityFactory)
    ClusterSharding(system).start(
      "MyProcessManager",
      Props(new MyProcessManager(factory)),
      ClusterShardingSettings(system),
      extractEntityId,
      extractShardId
    )
  }
  //#create-child-persistent-entity-factory

  private def extractEntityId: ShardRegion.ExtractEntityId = PartialFunction.empty
  private def extractShardId: ShardRegion.ExtractShardId = _ => ""
}

trait MyCommand

case object StartProcess extends MyCommand with ReplyType[String]

class MyProcessManagerEntity extends PersistentEntity {
  override type Command = MyCommand
  override type Event = String
  override type State = String
  override def initialState = ""
  override def behavior = {
    case "" => Actions()
  }
}

//#create-child-persistent-entity
class MyProcessManager(
    factory: ChildPersistentEntityFactory[MyProcessManagerEntity]
  ) extends Actor {

  private var entity: ChildPersistentEntity[MyCommand] = _

  override def preStart(): Unit = {
    entity = factory(
      context.self.path.name, // The entity id
      "entity"                // Name of the child actor
    )
  }
  //#create-child-persistent-entity


  override def receive = PartialFunction.empty

  private def demonstrateTell(): Unit = {
    //#child-persistent-entity-tell
    entity ! StartProcess
    //#child-persistent-entity-tell
  }

  private def demonstrateForward(): Unit = {
    //#child-persistent-entity-forward
    entity forward StartProcess
    //#child-persistent-entity-forward
  }

  private def demonstrateAsk(): Unit = {
    //#child-persistent-entity-ask
    import scala.concurrent.duration._
    implicit val timeout = Timeout(3.seconds)

    import context.dispatcher
    import akka.pattern.pipe

    val result = (entity ? StartProcess)
      .map(MappedReply.apply)

    result pipeTo self
    //#child-persistent-entity-ask
  }
}

case class MappedReply(reply: String)

// Ensures scalatest can't find the contained test
private abstract class Ignore {
  //#mock-child-persistent-entity
  class MyProcessManagerTest extends TestKit(ActorSystem())
    with WordSpecLike with BeforeAndAfterAll {

    override protected def afterAll(): Unit = {
      system.terminate()
    }

    "My process manager" should {
      "allow starting a process" in {
        val probe = TestProbe()

        val processManager = system.actorOf(Props(new MyProcessManager(
          // Mock the child persistent entity factory to use
          // our test probe.
          ChildPersistentEntityFactory.mocked[MyProcessManagerEntity](probe.testActor)
        )))

        // Send the process manager a start message
        processManager ! "start"
        // Expect the entity to receive a start process message
        probe.expectMsg(StartProcess)
        // Simulate the entity to reply with a started message
        probe reply "started"
        // Expect that message to be mapped forward back to us
        expectMsg(MappedReply("started"))

      }
    }
  }
  //#mock-child-persistent-entity
}
